AST Property tests

matheusaaguiar opened the pull request:
Suggested [here](https://github.com/ethereum/solidity/pull/13378#issuecomment-1298370722).
This adds a new kind of test which makes possible to check for specific properties of the AST instead of all its content.

matheusaaguiar committed:
Added new kind of test: check for specific properties of AST

nikola-matic commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,224 @@
+namespace
+{
+    using StringPair = std::pair<std::string, std::string>;
+    struct KeyValueParser
```
I'd argue this is a bit too much, especially given the test cases have a fairly simple pattern:
```
name: property.path.in.ast
```
which is basically just a `boost::algorithm::split()` on `:`, and then a `boost::algorithm::trim` on the resulting vector of string, asserting that its size is 2 (key, value), and that both elements pass the `isprint()` check.

matheusaaguiar replied:
Right. Agreed.

nikola-matic commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,224 @@
+                    nodesToVisit.push(node[memberName]);
+        }
+        else if (node.isArray())
+            for (
```
Pretty sure this could be a ranged for loop
```
for (auto&& member : node)
    nodesToVisit.push(member)
```

nikola-matic commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,224 @@
+    string documentation = "documentation";
+    string testCaseLine;
+
+    while(!nodesToVisit.empty())
```
The fact that we're resorting to an in-order queue based traversal of a JSON tree (I'd prefer a recursive traversal, but this is fine as well) is such an indictment against jsoncpp; with nlohman this would have been a single for loop over a `::flatten`ed JSON value :D

matheusaaguiar replied:
> with nlohman this would have been a single for loop over a ::flattened JSON value :D

That would be nice indeed. 

I had no particular reason for choosing this iterative approach over the recursive one. I didn't see any downside, so I just kept it.

nikola-matic commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,224 @@
+
+                    soltestAssert(!testCaseLine.empty());
+                    KeyValueParser parser{testCaseLine};
+                    auto [testId, testedProperty] = parser.parse()[0];
```
Asserting that whatever `parser.parse()` returns is not `empty()` would be preferable to a segfault.

nikola-matic commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,169 @@
+
+void ASTPropertyTest::readExpectations()
+{
+    for (auto [testId, testExpectation] : readKeyValuePairs(m_reader.simpleExpectations()))
```
```suggestion
    for (auto [testId, testExpectation]: readKeyValuePairs(m_reader.simpleExpectations()))
```

nikola-matic commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,169 @@
+{
+    if (!_property.empty())
+    {
+        string subNode = string(_property.substr(0, _property.find_first_of('.')));
```
```suggestion
        string subNode = _property.substr(0, _property.find_first_of('.'));
```

nikola-matic commented on test/libsolidity/ASTPropertyTests/nested_properties.sol:
```diff
@@ -0,0 +1,17 @@
+        for(uint i = 0; i < 42; ++i) {
+        }
+        /// TestCase2: initializationExpression.initialValue.value
+        for(uint i = 1; i < 42; i = i * 2) {
+         }
+         /// TestCase3: loopExpression.expression.subExpression.name
+         for(uint i = 0; i < 42; i++) {
+         }
```
Is this not indented properly to test `trim()`, or are you trying to summon Kamil?

matheusaaguiar replied:
Are you referring to the closing brackets being in a new line?

nikola-matic replied:
No, as in not not everything is in line, e.g. on line 8, you have an extra space.

matheusaaguiar replied:
Oh, right. I could not see that until now! :smile:

nikola-matic reviewed (dismissed):
Nitpicks only, looks good.

nikola-matic commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,169 @@
+    {
+        string subNode = _property.substr(0, _property.find_first_of('.'));
+        if (subNode != _property)
+            return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
```
Oooh, OK, my bad, you do need to explicitly create a `string` here - I didn't notice that `_property` was in fact a `string_view`.

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,169 @@
+        result.emplace_back(pair[0], pair[1]);
+    }
+    return result;
+}
```
I'm surprised that you even had to implement something like this here. I'd expect this to be already available somewhere. We already have expectations in this format in other test cases. For example in `functionSideEffects` or `gasTests`. We also have settings segment before expectations that has this format.

After checking if you could reuse what we already have I see that the problem is that the existing tests parse this as a part of a bigger format so they all implement it in a very non-reusable way. For example here's how we parse settings: https://github.com/ethereum/solidity/blob/b26090c2883a05fdfe2159746201087c454b57a8/test/TestCaseReader.cpp#L181-L191

I think it would be a good idea to make a small step towards making this more reusable though. This bit seems generic enough that I'd put it in `TestCaseReader`. Something like `expectationsAsKeyValuePairs()` that you would call instead of `simpleExpectations()`. Then we could later use it in other tests.

nikola-matic replied:
As long as `size_t colon = line.find(':')` is replaced with `boost::algorithm::split` :D

cameel commented on test/libsolidity/ASTPropertyTests/nested_properties.sol:
```diff
@@ -0,0 +1,17 @@
+        /// TestCase1: condition.operator
+        for(uint i = 0; i < 42; ++i) {
+        }
+        /// TestCase2: initializationExpression.initialValue.value
+        for(uint i = 1; i < 42; i = i * 2) {
+        }
+        /// TestCase3: loopExpression.expression.subExpression.name
+        for(uint i = 0; i < 42; i++) {
+        }
+    }
+}
+// ----
+// TestCase1: <
+// TestCase2: 1
+// TestCase3: i
```
I see a little problem with verbosity here. The way it is now you have to define a new ID for every property you want to test. So when writing such a test, I'll probably end up with something like this:
```solidity
/// MyLoopInitialValueProperty1: initializationExpression.initialValue.property1
/// MyLoopInitialValueProperty2: initializationExpression.initialValue.property2
/// MyLoopInitialValueProperty3: initializationExpression.initialValue.property3
for (...) { ... }
...
// ----
// MyLoopInitialValueProperty1: x
// MyLoopInitialValueProperty2: y
// MyLoopInitialValueProperty3: z
```

which is very verbose and will get annoying quickly.

IMO a much better way to do this would be:
```solidity
/// MyLoop
for (...) { ... }
...
// ----
// MyLoop.initialValue.property1: x
// MyLoop.initialValue.property2: y
// MyLoop.initialValue.property3: z
```

What do you think @ekpyron? I know you wanted to get this merged quickly not to delay the PR it depends on, but maybe we could at least consider it as a refactor once this gets merged?

nikola-matic replied:
For what it's worth, I think the current approach is more expressive, and more readable when looking at the test. On the other hand, it's a bit more annoying to write, but I think it's a fair price to pay.

cameel replied:
Hmm... why is it more expressive? It seems equivalent to me. I can't think of a case you would be able to do with one but not the other.

I also, don't think the more verbose version is more readable. It just has a lot of repetition that does not add anything of value. It requires you to invent more names, and these will either be short and hide detail or long and too verbose.

matheusaaguiar replied:
I was thinking about this before and was going towards something like a list of properties inside the Natspec comments... but after seeing what @cameel suggested, I think even that is unnecessary.
As in the example, the expectations themselves already request for the value of a given property. The test just need to search for the key-value pair in the corresponding annotated node.

ekpyron replied:
Just as a note on this: the current design is fine and we should merge it. We can still refactor it afterwards. The example you're giving, @cameel, is mostly skewed due to choosing long names. We'll likely have one test per property for which ``Case1`` is enough, after which the difference is negligible and it may even be preferable to have the subnode be part of the comment in code, since it doesn't change in expectation and thereby is closer to the place at which you need to trace the subpath.

So long story short, just keep it as is, we can still adjust, if it actually becomes a problem.

matheusaaguiar replied:
Another point to consider about design, but which could be done in the future, is regarding properties which are inside or are themselves an array of nodes.
For example, the name of the return variable of a function is inside property `returnParameters.parameters` which is a list. Another example, the name of parameters of a function is also inside a property which is a list.

cameel replied:
An easy way to handle that would be to allow expectation values to be arbitrary JSON. For easy parsing we would require them to be on a single line (i.e. no pretty-printing), which would also discourage trying to stuff things that are too complex in there. Would suffice for simple arrays at least. It's outside of the scope of this PR though.

cameel commented on test/libsolidity/ASTPropertyTests/simple_properties.sol:
`astPropertyTests` would be more consistent with the naming convention we have now. Most dirs are named that way and only `ASTJSON` and `ABIJson` stand out for whatever reason (and they're not even consistent with each other).

nikola-matic replied:
Yeah, the only reason I didn't mention this in my review is because of the ASTJSON tests, and I just assumed it's a precedent we establish for AST related tests; do you think renaming the ASTJSON tests to `astJson` would make sense?

cameel replied:
I agree that `astJson` and `abiJson` is what these should have been called and if they stay as they are now we'll just be getting more inconsistencies over time so nipping it in the bud is not a bad idea.

I wouldn't necessarily go out of my way to fix this, and definitely not in this PR, but I would also not reject a PR doing that if you really wanted to clean this up.

matheusaaguiar replied:
Fixing `ASTJSON` and `ABIJson` is a good idea, because indeed I chose to use upper case based on those names. 
I will change to `astPropertyTests`.

matheusaaguiar replied:
Renamed to `astPropertyTests`.

cameel reviewed (commented):
Here's some initial feedback.

I also pushed some tweaks as fixups. For example I noticed that the whole .cpp file was mixing tabs and spaces so I fixed that.

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,170 @@
+						docNode["text"].asString() :
+						docNode.asString();
+
+					soltestAssert(!testCaseLine.empty());
```
If you put more then one property test on a node, this code silently ignores that.

It probably also lets tests with duplicate IDs overwrite each other.

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,170 @@
+					auto propertyNode = findNode(node, testedProperty);
+					soltestAssert(propertyNode.has_value(), "Could not find "s + testedProperty);
+					soltestAssert(!propertyNode->isObject());
+					m_testCases[testId].obtainedValue = propertyNode->asString();
```
Most of these asserts have no messages. If you make a tiny mistake of any kind (missing colon, more than one colon, typo in property name, property being a dict), you'll just get a failed assert with no indication why.

This will be very frustrating to work with and it's not like it's that hard to add these messages. We should improve that before merging it.

cameel reviewed (commented):
This still needs some work. My comments from the previous review round were actually not addressed yet.

The biggest issue I see is the way errors are reported, i.e. when you make in the test syntax (see my comments below). I played with that a bit and added some messages that improve the situation, but there's more that should be done. I feel that there are still some easy ways to make a mistake that will result in an assert without any info what's wrong or even continue silently without error, so it will still be quite frustrating to work with.

r0qs commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,175 @@
+using namespace solidity::frontend;
+using namespace solidity::frontend::test;
+using namespace solidity;
+using namespace std;
```
We could already remove the `std` from here as well.

matheusaaguiar replied:
Done, but before committing, would like to know from @nikola-matic if it would be helpful or not, since he is doing that style change in the whole code base.

nikola-matic replied:
Please no, and thanks for asking 😅.

r0qs commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,175 @@
+		string subNode = string(_property.substr(0, _property.find_first_of('.')));
+		if (subNode != _property)
+			return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+		else if (_root.isMember(subNode))
```
```suggestion
		if (_root.isMember(subNode))
```

matheusaaguiar commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -110,38 +110,40 @@ void ASTPropertyTest::readTestedProperties(Json::Value const& _astJson)
+				auto const& [testId, testedProperty] = pairs[0];
+				m_testCases[testId].property = testedProperty;
+
+				soltestAssert(node.isMember("nodeType"));
```
I guess that this prevent that any metadata (?) node or something which is not AST node to be processed, but I am not sure if we need an informative error message here and how to phrase it.

cameel replied:
I did not add one here because if it fails, it's really a problem with the AST and not with the test. So it's instead something that would be a part of expectations if you care about the value of this field. This assert is here only because I included node type in the other assert below. Other that that, we would have been fine not checking for it.

But looking at it now, it would not hurt to have a message anyway. After all we're testing the AST and you could theoretically get a malformed one if you're changing code responsible for generating it. I would not expect this to happen very often, but still. You could use something like `"Invalid AST node: 'nodeType' field is missing."`.

matheusaaguiar reviewed (commented):
> This still needs some work. My comments from the previous review round were actually not addressed yet.
> The biggest issue I see is the way errors are reported

@cameel I think that I addressed all that was left, but one assert which I am not sure how to deal with (see bellow). 

Besides that, I allowed multiple tested properties per single node, which I guess it is something still a bit verbose, but offers us more flexibility while we decide and implement a nicer way as [discussed before](https://github.com/ethereum/solidity/pull/14309#discussion_r1237361089).

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,202 @@
+			pair.emplace_back(boost::trim_copy(element));
+		}
+
+		soltestAssert(pair.size() == 2, "Invalid property test: " + line);
```
These still need proper messages. The first one should say that the line contains non-printable characters.

The second should explicitly say that there's either too many or too few `:` on the line. Just saying that it's invalid is not enough IMO. There may be cases where the expectation  structured in such a way that it will not immediately jump at you that the extra colon is illegal. How about this for example?

```
// contract.fullyQualifiedName: "contract.sol:C"
```

Actually, now that I think of it, there's no reason not to accept an expectation like that. We should just split on the first colon and any remaining ones are a part of the value. We should only error out if there is no colon on the line.

matheusaaguiar replied:
Done.

cameel commented on test/libsolidity/ASTPropertyTest.h:
```diff
@@ -0,0 +1,69 @@
+	std::optional<Json::Value> findNode(Json::Value const& _root, std::string_view const& _property);
+
+	std::vector<std::string> m_expectationsSequence;
+	std::map<std::string, ASTPropertyTestCase> m_testCases;
```
I'd change naming here a bit:

```suggestion
	std::vector<std::string> m_testOrder;
	std::map<std::string, Test> m_tests;
```

I'm suggesting this now because I just saw `ASTPropertyTestCase` in the code and for a moment got very confused, remembering that we use the `XYZTestCase` naming for classes derived from `TestCase`. So having the struct called like that is a bit misleading. My suggestion would be to simply call the struct `Test` (it's local to the class anyway) and have the whole class named `ASTPropertyTestCase` instead.

cameel replied:
Also, how about renaming `readTestedProperties()` to `extractTestsFromAST()`?
Also `generateTestCaseValues()` -> `formatExpectations()`.

matheusaaguiar replied:
I only didn't rename `ASTPropertyTestCase`, which I would prefer to keep as is. I checked and there's only class `EVMVersionRestrictedTestCase` derived from `TestCase` which uses the suggested naming scheme.

cameel commented on test/libsolidity/ASTPropertyTest.h:
```diff
@@ -0,0 +1,69 @@
+private:
+	struct ASTPropertyTestCase
+	{
+		std::string testIdentifier;
```
This field seems unused. You're always using the map key instead.

cameel commented on test/libsolidity/ASTPropertyTest.h:
```diff
@@ -0,0 +1,69 @@
+	std::vector<StringPair> readKeyValuePairs(std::string const& _input);
+	void readTestedProperties(Json::Value const& _astJson);
+	void generateTestCaseValues(std::string& _values, bool _obtained = true);
+	std::optional<Json::Value> findNode(Json::Value const& _root, std::string_view const& _property);
```
I think it would be cleaner for `generateTestCaseValues()` to return a value instead of having a mutable parameter.

Also, `string_view` is typically passed by value.

```suggestion
	std::string generateTestCaseValues(bool _obtained = true);
	std::optional<Json::Value> findNode(Json::Value const& _root, std::string_view _property);
```

Finally, at the point of call it's nice to add a comment indicating the name of the parameter if the value passed in does not make that clear:
```c++
generateTestCaseValues(false /* _obtained */);
```

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,202 @@
+	generateTestCaseValues(m_expectation, false);
+}
+
+optional<Json::Value> ASTPropertyTest::findNode(Json::Value const& _root, string_view const& _property)
+{
+	if (_root.isObject() && !_property.empty())
+	{
+		string subNode = string(_property.substr(0, _property.find_first_of('.')));
+		if (subNode != _property)
+			return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+		if (_root.isMember(subNode))
+			return make_optional(_root[subNode]);
+	}
+	return {};
+}
```
Is this really correct? I see that when `subNode != _property` you use `_root[subNode]` without checking if it's present.

I'd also adjust naming and restructure it a bit. How about this instead?

```suggestion
optional<Json::Value> ASTPropertyTest::jsonValueByPath(Json::Value const& _node, string_view _jsonPath)
{
	if (!_node.isObject() || _jsonPath.empty())
		return {};

	string_view memberName = _jsonPath.substr(0, _jsonPath.find_first_of('.'));
	if (!_node.isMember(memberName))
		return {};
	if (memberName == _jsonPath)
		return _node[memberName];

	return findNode(_node[memberName], _jsonPath.substr(memberName.size() + 1));
}
```

Also, this seems like a pretty generic helper that might also be helpful outside of tests. Why not put it in `libsolutil/JSON.h`?

matheusaaguiar replied:
> Is this really correct? I see that when subNode != _property you use _root[subNode] without checking if it's present.

In fact, it is not. If `subNode` is not present, operator `[]` returns a `null` value, which then I assumed that `isObject` would return `false`, however seems that's not the [case](https://jsoncpp.sourceforge.net/json__value_8cpp_source.html#l01431).
I will refactor as you suggested. 

> Also, this seems like a pretty generic helper that might also be helpful outside of tests. Why not put it in libsolutil/JSON.h?

Yeah, true. That's a good idea. I even searched for something like this before implementing it.

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,202 @@
+				string nodeDocstring = node["documentation"].isObject() ?
+					node["documentation"]["text"].asString() :
+					node["documentation"].asString();
+				soltestAssert(!nodeDocstring.empty());
```
Can this actually happen in valid code? For example, would such code trigger this?
```solidity
///
contract C {}
```
or this code?
```solidity
/***/
contract C {}
```

Or does the text always include the comment markers?

If it can happen, then this should not really be an assert. We should instead `continue` instead of failing. If that's the case, please also include a situation like this in one of your test cases.

matheusaaguiar replied:
No, it can't actually happen. Empty `natspec` documentation is not generated.

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,202 @@
+				soltestAssert(!nodeDocstring.empty());
+
+				vector<StringPair> pairs = readKeyValuePairs(nodeDocstring);
+				soltestAssert(!pairs.empty(), "It is required at least one property test per annotated AST node.");
```
I think that not specifying any tests should be allowed.

Actually, I just realized that this mechanism does not really allow testing code with Natspec because we assume Natspec always contains property tests. I guess there's not much we can do about that without complicating the implementation so it'll have to stay that way, but still, I'm a bit surprised by this.

matheusaaguiar replied:
This assert will catch the case when there is an empty line (actually a line with spaces only) between two valid test lines:
```
/// Test1: name
///    
/// Test2: visibility
```
If the line is really empty, i.e, no spaces at all, then it doesn't show in the ast JSON.

cameel replied:
Hmm... why would that check catch the case you're showing? I'd expect `pairs.empty()` to be true because you'll still have `Test1: name` and `Test2: visibility` in the vector.

matheusaaguiar replied:
Yeah, true. I guess I was thinking of something else when I wrote that... :smile: 
It will actually catch the case when all of the lines are blank (filled only with spaces).

matheusaaguiar replied:
We could also just ignore and continue processing... but would it make sense to have a blank test?

cameel replied:
It would not be terribly useful, but I don't think it should be an error. I actually think it would be marginally useful to even have non-empty Natspec comments that are not tests, but that's a bigger change and not worth it in terms of effort. Ignoring empty ones is something we can very easily do though.

matheusaaguiar replied:
Ok. It now ignores empty/blank tests, i.e, all lines are empty or have only whitespaces.

cameel replied:
Ah, I just realized what you were worried about. I see the test with no expectations at all :)

Well, you should be able to have empty Natspec lines, but it would still be useful to enforce that the file as a whole has at least one expectation defined. That's what we do for example in semantic tests. Could you add an assert at the end of the function, checking that there is at least one expectation line at the end?

Also, would be helpful to assert that all defined properties are actually used in expectations. If not, the person writing the case most likely forgot one of them.

matheusaaguiar replied:
> Could you add an assert at the end of the function, checking that there is at least one expectation line at the end?

Done.

> Also, would be helpful to assert that all defined properties are actually used in expectations. If not, the person writing the case most likely forgot one of them.

That was already implemented. Any tests with no corresponding property or expectation produce an assertion fail.

cameel replied:
> That was already implemented. Any tests with no corresponding property or expectation produce an assertion fail.

I meant checking at there's at least one test present below `// ----`.

cameel replied:
> That was already implemented. Any tests with no corresponding property or expectation produce an assertion fail.

Ah, right. The `does not have a corresponding expected value` check does cover that.

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,202 @@
+
+				for (auto const& [testId, testedProperty]: pairs)
+				{
+					soltestAssert(
+						readTests.insert(testId).second,
+						"Test \"" + testId + "\" was already defined before."
+					);
```
I'd avoid asserts with side-effects as a general rule.

It's less of a problem with our asserts than with the standard `assert()` macro, since our asserts are always executed, but still, it feels just wrong and while reading code I rarely expect such code to actually do anything other than a simple side-effect-free check.
```suggestion
					bool inserted = readTests.insert(testId).second;
					soltestAssert(inserted,	"Test \"" + testId + "\" is already defined.");
```

cameel replied:
By the way, `readTests` actually seems a bit superfluous. You could disallow tests with empty property path (they are not useful anyway) and determine if the test has been defined by checking if `property` field is empty.

matheusaaguiar replied:
Yeah, I removed `readTests` and now the code checks `property` field as you suggested.

cameel commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,202 @@
+	compiler.setEVMVersion(solidity::test::CommonOptions::get().evmVersion());
+	compiler.setOptimiserSettings(solidity::test::CommonOptions::get().optimize);
+	if (!compiler.parseAndAnalyze())
+		BOOST_THROW_EXCEPTION(runtime_error("Parsing contract failed"));
```
If you do this, the only thing you'll see when the test has compilation errors will be `Parsing contract failed`, without any details about where the error is.

Please do at least this:
```suggestion
		BOOST_THROW_EXCEPTION(runtime_error(
			"Parsing contract failed:\n" +
			SourceReferenceFormatter::formatErrorInformation(compiler.errors(), compiler, _formatted)
		));
```

Ideally, your test class would inherit from `AnalysisFramework` instead and use its compiler stack and error formatting functions. Though for now just adding errors to the message will be fine. We can refactor this later.

matheusaaguiar commented on test/libsolidity/ASTPropertyTest.cpp:
```diff
@@ -0,0 +1,202 @@
+	vector<StringPair> result;
+	for (string const& inputLine: _input | ranges::views::split('\n') | ranges::to<vector<string>>)
+	{
+		string line = boost::trim_copy(inputLine);
```
This avoids having a line with only spaces being processed further.
I was trying to do this as a step of the `ranges` pipeline in the `for` loop above, but I'll have to learn how to do that first :sweat_smile:

cameel replied:
I don't think `ranges` has `trim()` so this is the way to do it currently :)

cameel replied:
Actually, maybe you could use `transform()` with `trim_copy` as the function. This will still probably end up less concise than this.

One thing you could change though would be make `inputLine` a `string` (rather than a reference) and then use normal `trim()` in place. `to<>` has to create a copy anyway. But that's just a microoptimization :)

matheusaaguiar replied:
> Actually, maybe you could use transform() with trim_copy as the function. This will still probably end up less concise than this.

I looked at some examples and docs and was considering using `transform`, but from what I learned, I also think it would end up being less concise.

> One thing you could change though would be make inputLine a string [...]

I will change that. It's fairly simple.

matheusaaguiar reviewed (commented):
Rebased on develop. Squashed some commits with cleanups and refactors.

cameel reviewed (commented):
I rechecked the PR and it seems almost ready now. The only remaining (minor) issue is https://github.com/ethereum/solidity/pull/14309#discussion_r1291765856.

There were some small tweaks needed in some error messages and names but I applied them myself (see the fixups I pushed).

cameel commented:
So, https://github.com/ethereum/solidity/pull/14309#discussion_r1291765856 is still the last remaining thing to do here. Once you address it, you can squash the commits into some reasonable and we could merge.

One more thing though - due to @nikola-matic's purge, you have to deal with `namespace std` in your PR first :)

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `554874f2e8fee72aecb205604f7cd49a78abeb94`:
```
Error: Trailing whitespace found:
test/libsolidity/astPropertyTests/blank_test_case.sol:3:    ///    
test/libsolidity/astPropertyTests/blank_test_case.sol:4:    /// 
test/libsolidity/astPropertyTests/blank_test_case.sol:5:    ///  
test/libsolidity/astPropertyTests/blank_test_case.sol:9:// 
test/libsolidity/astPropertyTests/blank_test_case.sol:10:// 
test/libsolidity/astPropertyTests/blank_test_case.sol:11:// 

```
Please check that your changes are working as intended.

matheusaaguiar commented:
> So, [#14309 (comment)](https://github.com/ethereum/solidity/pull/14309#discussion_r1291765856) is still the last remaining thing to do here. Once you address it, you can squash the commits into something more reasonable and we could merge.

Done. Squashed and rebased.
> 
> There's one complication though - due to @nikola-matic's purge, you have to deal with `namespace std` in your PR first :)

Thankfully, it was just a single function that needed adjustments :relieved:
